src/untyped_array.rs

```diff
@@ -246,12 +244,233 @@ impl PyUntypedArray {
+    ///
+    /// ```
+    /// use numpy::PyArray3;
+    /// use pyo3::Python;
+    ///
+    /// Python::with_gil(|py| {
+    ///     let arr = PyArray3::<f64>::zeros(py, [4, 5, 6], false);
+    ///
+    ///     assert_eq!(arr.ndim(), 3);
+    /// });
+    /// ```
+    ///
+    /// [ndarray-ndim]: https://numpy.org/doc/stable/reference/generated/numpy.ndarray.ndim.html
+    /// [PyArray_NDIM]: https://numpy.org/doc/stable/reference/c-api/array.html#c.PyArray_NDIM
+    #[inline]
+    fn ndim(&self) -> usize {
+        unsafe { (*self.as_array_ptr()).nd as usize }
+    }
+
+    /// Returns a slice indicating how many bytes to advance when iterating along each axis.
+    ///
+    /// See also [`ndarray.strides`][ndarray-strides] and [`PyArray_STRIDES`][PyArray_STRIDES].
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use numpy::PyArray3;
+    /// use pyo3::Python;
+    ///
+    /// Python::with_gil(|py| {
+    ///     let arr = PyArray3::<f64>::zeros(py, [4, 5, 6], false);
+    ///
+    ///     assert_eq!(arr.strides(), &[240, 48, 8]);
+    /// });
+    /// ```
+    /// [ndarray-strides]: https://numpy.org/doc/stable/reference/generated/numpy.ndarray.strides.html
+    /// [PyArray_STRIDES]: https://numpy.org/doc/stable/reference/c-api/array.html#c.PyArray_STRIDES
+    #[inline]
+    fn strides(&self) -> &[isize] {
+        let n = self.ndim();
+        if n == 0 {
+            cold();
+            return &[];
+        }
+        let ptr = self.as_array_ptr();
+        unsafe {
+            let p = (*ptr).strides;
+            slice::from_raw_parts(p, n)
+        }
+    }
+
+    /// Returns a slice which contains dimmensions of the array.
+    ///
+    /// See also [`ndarray.shape`][ndaray-shape] and [`PyArray_DIMS`][PyArray_DIMS].
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use numpy::PyArray3;
+    /// use pyo3::Python;
+    ///
+    /// Python::with_gil(|py| {
+    ///     let arr = PyArray3::<f64>::zeros(py, [4, 5, 6], false);
+    ///
+    ///     assert_eq!(arr.shape(), &[4, 5, 6]);
+    /// });
+    /// ```
+    ///
+    /// [ndarray-shape]: https://numpy.org/doc/stable/reference/generated/numpy.ndarray.shape.html
+    /// [PyArray_DIMS]: https://numpy.org/doc/stable/reference/c-api/array.html#c.PyArray_DIMS
+    #[inline]
+    fn shape(&self) -> &[usize] {
+        let n = self.ndim();
+        if n == 0 {
+            cold();
+            return &[];
+        }
+        let ptr = self.as_array_ptr();
+        unsafe {
+            let p = (*ptr).dimensions as *mut usize;
+            slice::from_raw_parts(p, n)
+        }
+    }
+
+    /// Calculates the total number of elements in the array.
+    fn len(&self) -> usize {
+        self.shape().iter().product()
+    }
+
+    /// Returns `true` if the there are no elements in the array.
+    fn is_empty(&self) -> bool {
         self.shape().iter().any(|dim| *dim == 0)
     }
 }
+
+fn check_flags(bound: &Bound<'_, PyUntypedArray>, flags: i32) -> bool {
```

Comment:
Similarly, if this takes an `*mut npyffi::PyArrayObject` directly, the three flag-checking method could be implemented directly on the trait.